pkg/util/cache/cache.go

```diff
@@ -68,6 +80,123 @@ func NewLRUCache[K comparable, V any](
 	}
 }
 
+func (c *lruCache[K, V]) TryPin(keys ...K) (items []*cacheItem[K, V], ok bool) {
+	c.rwlock.Lock()
+	defer c.rwlock.Unlock()
+
+	var notExisted int32
+	for _, k := range keys {
+		iter, ok := c.items[k]
+		if ok {
+			item := iter.Value.(*cacheItem[K, V])
+			item.pinCount.Inc()
+		} else {
+			notExisted++
+		}
+	}
+	defer func() {
+		if !ok {
+			for _, k := range keys {
+				iter, ok2 := c.items[k]
+				if ok2 {
+					item := iter.Value.(*cacheItem[K, V])
+					item.pinCount.Dec()
+				}
+			}
+		}
+	}()
+
+	if c.size+notExisted > c.cap {
+		if ok = c.tryEvict(notExisted); !ok {
+			return nil, false
+		}
+	}
+
+	ret := make([]*cacheItem[K, V], 0, len(keys))
+	for _, key := range keys {
+		if iter, ok := c.items[key]; ok {
+			item := iter.Value.(*cacheItem[K, V])
+			c.accessList.Remove(iter)
+			c.accessList.PushFront(item)
+			ret = append(ret, item)
+		} else {
+			item := newCacheItemWithKey[K, V](key)
+			item.pinCount.Inc()
+			ret = append(ret, item)
+			c.add(item)
+		}
+	}
+	return ret, true
+}
+
+func (c *lruCache[K, V]) tryEvict(count int32) bool {
+	item := c.accessList.Back()
+	var n int32
+	evictItemsKey := make([]K, 0, count)
+	for item != nil && n < count {
+		cItem := item.Value.(*cacheItem[K, V])
+		if cItem.pinCount.Load() == 0 {
```

Comment:
Provided that the least recent used item has a reference count > 0, is there any other item evictable? Can we just iterate next instead of returning false here?